parser: logfmt: Allow the logfmt parser to reject messages that have keys without values

seveas opened the pull request:
To allow a gradual move towards logfmt and otel messages, we need to enable logfmt parsing in a way that does not try to parse non-logfmt messages. Unlike the json parser, which has an easy time rejecting non-json messages, the logfmt parser will hapily accept any log messages and parse them into something quite unusable.

By making it possible (but not required) that all items in a logfmt message have a value, we give the logfmt parser the capability to reject message and enable a gradual move towards logfmt messages.

We've been running with this (in a different form, as a parser that always rejects lines with value-les kesy) in production for over a year, and I can say it's been the only way to get people to adopt structured logging. In an environment where random libraries can throw things into your logstream, it's just been impossible to rely on everything being properly logfmt'ed.

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [ ] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [ ] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

seveas commented:
Example config:

```
[INPUT]
    name tail
    path /var/log/messy-application.log
    parser logfmt_sloppy
```

```
[Parser]
    Name logfmt_sloppy
    Format logfmt
    No_Bare_Keys true
```

seveas commented:
Debug logs using this logfile:

```
foo=bar baz=quux plonk1=1 plonk2=2 last_plonk=last
foo=bar baz=quux plonk1 plonk2 last_plonk
```

```
Fluent Bit v2.0.0
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/09/26 15:37:33] [ info] Configuration:
[2022/09/26 15:37:33] [ info]  flush time     | 1.000000 seconds
[2022/09/26 15:37:33] [ info]  grace          | 5 seconds
[2022/09/26 15:37:33] [ info]  daemon         | 0
[2022/09/26 15:37:33] [ info] ___________
[2022/09/26 15:37:33] [ info]  inputs:
[2022/09/26 15:37:33] [ info]      tail
[2022/09/26 15:37:33] [ info] ___________
[2022/09/26 15:37:33] [ info]  filters:
[2022/09/26 15:37:33] [ info] ___________
[2022/09/26 15:37:33] [ info]  outputs:
[2022/09/26 15:37:33] [ info]      stdout.0
[2022/09/26 15:37:33] [ info] ___________
[2022/09/26 15:37:33] [ info]  collectors:
[2022/09/26 15:37:33] [ info] [fluent bit] version=2.0.0, commit=60b5d67afd, pid=9783
[2022/09/26 15:37:33] [debug] [engine] coroutine stack size: 12288 bytes (12.0K)
[2022/09/26 15:37:33] [ info] [storage] version=1.2.0, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/09/26 15:37:33] [ info] [cmetrics] version=0.4.0
[2022/09/26 15:37:33] [debug] [tail:tail.0] created event channels: read=21 write=22
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] flb_tail_fs_stat_init() initializing stat tail input
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] scanning path last-item.log
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] inode=116574479 with offset=0 appended as last-item.log
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] scan_glob add(): last-item.log, inode 116574479
[0] tail.0: [1664199453.987871000, {"foo"=>"bar", "baz"=>"quux", "plonk1"=>"1", "plonk2"=>"2", "last_plonk"=>"last"}]
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] 1 new files found on path 'last-item.log'
[2022/09/26 15:37:33] [ info] [output:stdout:stdout.0] worker #0 started
[2022/09/26 15:37:33] [debug] [stdout:stdout.0] created event channels: read=28 write=29
[1] tail.0: [1664199453.987877000, {"log"=>"foo=bar baz=quux plonk1 plonk2 last_plonk"}]
[2022/09/26 15:37:33] [debug] [router] match rule tail.0:stdout.0
[2022/09/26 15:37:33] [ info] [http_server] listen iface=127.0.0.1 tcp_port=2020
[2022/09/26 15:37:33] [ info] [sp] stream processor started
[2022/09/26 15:37:33] [debug] [input chunk] update output instances with new chunk size diff=122
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] [static files] processed 93b
[2022/09/26 15:37:34] [debug] [out flush] cb_destroy coro_id=0
[2022/09/26 15:37:33] [ info] [input:tail:tail.0] inode=116574479 file=last-item.log ended, stop
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] inode=116574479 file=last-item.log promote to TAIL_EVENT
[2022/09/26 15:37:33] [debug] [input:tail:tail.0] [static files] processed 0b, done
[2022/09/26 15:37:33] [debug] [task] created task=0x60000123c070 id=0 OK
[2022/09/26 15:37:33] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2022/09/26 15:37:33] [ warn] [engine] service will shutdown in max 5 seconds
[2022/09/26 15:37:33] [ info] [input] pausing tail.0
[2022/09/26 15:37:34] [debug] [task] destroy task=0x60000123c070 (task_id=0)
[2022/09/26 15:37:34] [ info] [engine] service has stopped (0 pending tasks)
[2022/09/26 15:37:34] [ info] [input] pausing tail.0
[2022/09/26 15:37:34] [debug] [input:tail:tail.0] inode=116574479 removing file name last-item.log
[2022/09/26 15:37:34] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/09/26 15:37:34] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

seveas commented:
local-build-all.sh fails with an error unrelated to this change:

```
#11 2.150 CMake Error at lib/cfl/src/CMakeLists.txt:24 (install):
#11 2.150   install TARGETS given no ARCHIVE DESTINATION for static library target
#11 2.150   "cfl-static".
#11 2.150
#11 2.150
```

nokute78 commented on include/fluent-bit/flb_parser.h:
```diff
@@ -50,6 +50,7 @@ struct flb_parser {
     int time_offset;      /* fixed UTC offset */
     int time_keep;        /* keep time field */
     int time_strict;      /* parse time field strictly */
+    int no_bare_keys;     /* in logfmt parsers, require all keys to have values */
```
Could you rename like `logfmt_no_bare_keys` ?
This parameter is specific for logfmt.

nokute78 reviewed (changes requested):
- Could you fix test code build error ?
  - https://github.com/fluent/fluent-bit/actions/runs/3128058880/jobs/5144817204#step:5:2383
  - You can build test code `cmake .. -DFLB_TESTS_RUNTIME=on -DFLB_TESTS_INTERNAL=on && make`
- Could you modify commit message ?
  - https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#commit-changes

seveas commented:
I've addressed all the feedback. Or at least I think I have, running the tests is a pain on macos due to the plethora of firewall popups, so I'm hoping CI will confirm that I've fixed it all 😀

nokute78 commented:
@seveas Could you check below error ?
https://github.com/fluent/fluent-bit/actions/runs/3172866465/jobs/5203552550#step:4:2230

https://github.com/seveas/fluent-bit/blob/logfmt-strict/tests/internal/fuzzers/parse_logfmt_fuzzer.c#L20

seveas commented:
CI is happy, and I created a docs PR. I think this is ready for re-review.

edsiper commented:
@seveas can you pls rebase on top of master ? (there is a conflict due to recent changes)

seveas commented:
It's been rebased and the (fortunately simple) conflict is gone.

nokute78 reviewed (approved):
LGTM

seveas commented:
Ping. It's almost 2 months since the last comment on this pr. Would love to get this into the next version.
